analyzers/src/SonarAnalyzer.Common/Rules/Utilities/UtilityAnalyzerBase.cs

```diff
@@ -87,40 +88,49 @@ public abstract class UtilityAnalyzerBase<TSyntaxKind, TMessage> : UtilityAnalyz
         protected UtilityAnalyzerBase(string diagnosticId, string title) : base(diagnosticId, title) { }
 
         protected sealed override void Initialize(SonarAnalysisContext context) =>
-            context.RegisterCompilationAction(c =>
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                ReadParameters(startContext);
+                if (!IsAnalyzerEnabled)
+                {
+                    return;
+                }
+
+                var treeMessages = new List<TMessage>();
+                startContext.RegisterSemanticModelAction(modelContext =>
                 {
-                    ReadParameters(c);
-                    if (!IsAnalyzerEnabled)
+                    if (ShouldGenerateMetrics(modelContext))
                     {
-                        return;
+                        treeMessages.Add(CreateMessage(modelContext.Tree, modelContext.SemanticModel));
                     }
+                });
 
-                    var treeMessages = c.Compilation.SyntaxTrees
-                        .Where(x => ShouldGenerateMetrics(c, x))
-                        .Select(x => CreateMessage(x, c.Compilation.GetSemanticModel(x)));
-                    var messages = CreateAnalysisMessages(c)
+                startContext.RegisterCompilationEndAction(endContext =>
+                {
+                    var allMessages = CreateAnalysisMessages(endContext)
                         .Concat(treeMessages)
                         .WhereNotNull()
                         .ToArray();
                     lock (FileWriteLock)
                     {
                         Directory.CreateDirectory(OutPath);
                         using var stream = File.Create(Path.Combine(OutPath, FileName));
-                        foreach (var message in messages)
+                        foreach (var message in allMessages)
                         {
                             message.WriteDelimitedTo(stream);
                         }
                     }
                 });
+            });
 
         protected virtual bool ShouldGenerateMetrics(SyntaxTree tree) =>
             // The results of Metrics and CopyPasteToken analyzers are not needed for Test projects yet the plugin side expects the protobuf files, so we create empty ones.
             (AnalyzeTestProjects || !IsTestProject)
             && FileExtensionWhitelist.Contains(Path.GetExtension(tree.FilePath))
             && (AnalyzeGeneratedCode || !Language.GeneratedCodeRecognizer.IsGenerated(tree));
 
-        private bool ShouldGenerateMetrics(SonarCompilationReportingContext context, SyntaxTree tree) =>
-            (AnalyzeUnchangedFiles || !context.IsUnchanged(tree))
-            && ShouldGenerateMetrics(tree);
+        private bool ShouldGenerateMetrics<T>(SonarAnalysisContextBase<T> context) =>
```

Comment:
This is actually wrong. It needs to be reverted to the previous version with `SonarSematicModelReportingContext`.

While it compiles for all of them, generic `ShouldGenerateMetrics` would provide wrong result for some of the classes that inherit it. That's why `context.IsUnchanged` has explicit `tree` parameter and do not take `context.Tree` inside it.
```suggestion
        private bool ShouldGenerateMetrics(SonarSematicModelReportingContext context) =>
```

Reply:
Once it's `SonarSematicModelReportingContext`, then using `context.Tree` is valid

Reply:
I don't get why it isn't okay to have `SonarAnalysisContextBase<T>` as a parameter as it is the base class of SonarSematicModelReportingContext.

Given that this is the base class for all utility analyzers I would expect this class to also allow to register for SyntaxNode actions, code block actions, symbol actions, and so on to collect the messages. `SonarAnalysisContextBase<T>` or even `SonarAnalysisContextBase` should therefore be used here as parameter type IMO.

Reply:
The reason is that the `SonarAnalysisContextBase<T>` contract satisfies also other contexts, including problematic 
https://github.com/SonarSource/sonar-dotnet/blob/8f653cd42341dae603320cecae1148ea5d77b1f1/analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarSymbolReportingContext.cs#L25
and
https://github.com/SonarSource/sonar-dotnet/blob/8f653cd42341dae603320cecae1148ea5d77b1f1/analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationReportingContext.cs#L32

It would be better if the `Tree` wasn't in that base class but in `SonarTreeReportingContextBase` instead. There are no good reasons for this, only legacy (bad) reason. I've created #6729 to fix this later (I'll add it to my notes to fix sooner than later)